src/server/domain/consents/{ID}/generateChallenge.ts

```diff
@@ -0,0 +1,100 @@
+ are distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ KIND, either express or implied. See the License for the specific language
+ governing permissions and limitations under the License.
+ Contributors
+ --------------
+ This is the official list of the Mojaloop project contributors for this file.
+ Names of the original copyright holders (individuals or organizations)
+ should be listed with a '*' in the first column. People who have
+ contributed from an organization can be listed under the organization
+ that actually holds the copyright for their contributions (see the
+ Gates Foundation organization for an example). Those individuals should have
+ their names indented and be marked with a '-'. Email address can be added
+ optionally within square brackets <email>.
+ * Gates Foundation
+ - Name Surname <name.surname@gatesfoundation.com>
+
+ - Abhimanyu Kapur <abhi.kapur09@gmail.com>
+
+ --------------
+ ******/
+
+import { Request } from '@hapi/hapi'
+import { consentDB } from '../../../../lib/db'
+import { thirdPartyRequest } from '../../../../lib/requests'
+import { Consent } from '../../../../model/consent'
+import { Enum } from '@mojaloop/central-services-shared'
+import SDKStandardComponents from '@mojaloop/sdk-standard-components'
+
+/**
+ * Validates whether generate challenge request is valid
+ * by comparing consent ID sent matches with existing consent in table
+ * and if source ID matches initiator ID
+ * @param request: request received from PISP
+ * @param consent: Consent object
+ */
+export function isConsentRequestValid (
+  request: Request,
+  consent: Consent): boolean {
+  const fspiopSource = request.headers[Enum.Http.Headers.FSPIOP.SOURCE]
+  return (consent && consent.initiatorId === fspiopSource)
+}
+
+/**
+ * Assigns credentials to given consent object and updates in the database
+ * Returns updated consent object
+ */
+export async function updateCredential (
+  consent: Consent,
+  challenge: string,
+  credentialType: string,
+  credentialStatus: string): Promise<Consent> {
+  // Update consent credentials
+  consent.credentialType = credentialType
+  consent.credentialStatus = credentialStatus
+  consent.credentialChallenge = challenge
+
+  // Update in database
+  await consentDB.update(consent)
+  return consent
+}
+
+/**
+ * Builds body of outgoing request and makes PUT consents/{ID} call to server
+ * @param consent Consent object with credential challenge, type and status
+ * @param headers headers from PISP generate challenge request
+ */
+export async function putConsentId (
+  consent: Consent,
+  request: Request):
+  Promise<SDKStandardComponents.GenericRequestResponse | undefined> {
+  // Construct body of outgoing request
+  const body = {
+    requestId: consent.id,
+    initiatorId: consent.initiatorId as string,
+    participantId: consent.participantId as string,
+    scopes,
+    credential: {
+      id: null,
```

Comment:
why `null` here

Reply:
this is based on the linking diagram

Reply:
https://camo.githubusercontent.com/57f9a75d95d5bdd24f7563ce5c81579c8aa9db24/687474703a2f2f7777772e706c616e74756d6c2e636f6d2f706c616e74756d6c2f70726f78793f63616368653d6e6f267372633d68747470733a2f2f7261772e67697468756275736572636f6e74656e742e636f6d2f6d6f6a616c6f6f702f706973702f6d61737465722f646f63732f6c696e6b696e672f35612d63726564656e7469616c2d726567697374726174696f6e2e70756d6c